Foundation/String/UTF8.hs

```diff
@@ -820,6 +822,76 @@ reverse s@(String ba) = runST $ do
                 _  -> return () -- impossible
             loop ms (si `offsetPlusE` nb) d
 
+-- Finds how many times we need to perform the replacement, and where are
+-- the insertion points.
+indices :: String -> String -> (Int, [Offset8])
+indices needle@(String ned) haystack@(String hy) =
```

Comment:
Suggest you generalise this to UArray and implement the String one via newtype wrapping on that.